File: js/test/htdocs/spec/test_protocol_timeout.js

describe('Protocol (Timeout)', function() {

    beforeEach(() => {
        jasmine.clock().install();
    });
    
    afterEach(() => {
        jasmine.clock().uninstall();
    });
    
    it('Timeout (Receive)', async function() {
		const test = new TestProtocolTimeout();
        await test.testReceiveTimeout();
    });    

    it('Timeout (Send)', async function() {
		const test = new TestProtocolTimeout();
        await test.testSendTimeout();
    });    
});


class TestProtocolTimeout {	
	
	async testReceiveTimeout() {
        jasmine.clock().mockDate(new Date(1000));

        // Bridge for sending data
        const bridgeSend = new JsMidiBridge();
        
        bridgeSend.getFile = function(path_p) {
			if (path_p == "foo") {
				return "OllehCasklakldmcömpöosdpo sdvsdöv sdvsd vsdvgeg etrhtrhrt ret";
			}
			if (path_p == "bar") {
				return "Hello sdv wrpv wprvmevmkefv ö löevö wöorvömklfsdvnlsd dvsdvsdvsdvsssvsd";
			}						
		}
		
        // Bridge for receiving data
        const bridgeReceive = new JsMidiBridge();

		let finishCalls = [];
		
		bridgeReceive.onReceiveFinish = function(data) {
			finishCalls.push(data.data);
		}
		
		let messagesSentSend = [];
		bridgeSend.sendSysex = function(manufacturerId, data) {
			messagesSentSend.push({
                manufacturerId: manufacturerId,
                data: data
            });
		}

        let messagesSentReceive = [];
		bridgeReceive.sendSysex = function(manufacturerId, data) {
			messagesSentReceive.push({
                manufacturerId: manufacturerId,
                data: data
            });
		}	

        // Get a request message for the path from the receiving side
        await bridgeReceive.request("foo", 3);
        const msgRequestFoo = messagesSentReceive[0];
        messagesSentReceive = [];

        // Get a request message for the path from the receiving side
        await bridgeReceive.request("bar", 4);
        const msgRequestBar = messagesSentReceive[0];
        messagesSentReceive = [];

        // Let the bridge receive a request message, to trigger it sending a file
        expect(await bridgeSend.receive(msgRequestFoo)).toBe(true);

        // Receive start message
        const startMsg1 = messagesSentSend.shift();
        const dataMsg1 = messagesSentSend.shift();
        expect(await bridgeReceive.receive(startMsg1)).toBe(true);
        expect(messagesSentReceive.length).toBe(0);

        // Receive data message
        expect(await bridgeReceive.receive(dataMsg1)).toBe(true);

        // Get next data message
        expect(messagesSentReceive.length).toBe(1);
        const ackMsg1 = messagesSentReceive.shift();
        expect(await bridgeSend.receive(ackMsg1)).toBe(true);
        const dataMsg1b = messagesSentSend.shift();        
        
        // Let pass some virtual time so the first transmission will timeout
        jasmine.clock().mockDate(new Date(1000 * 10));
        
        // Let the sending bridge receive a request message, to trigger it sending a file
        expect(await bridgeSend.receive(msgRequestBar)).toBe(true);
        
        // Receive start message 
        const startMsg2 = messagesSentSend.shift();
        const dataMsg2 = messagesSentSend.shift();
        expect(await bridgeReceive.receive(startMsg2)).toBe(true);
        expect(messagesSentReceive.length).toBe(0);

        // Receive data message 2 (must raise "Transmission not found")
        expect(await bridgeReceive.receive(dataMsg1b)).toBe(true);

        await this.#evaluateError(bridgeReceive, messagesSentReceive, ["transmission", "not found"]);
    }

    async testSendTimeout() {
        jasmine.clock().mockDate(new Date(1000));

        // Bridge for sending data
        const bridgeSend = new JsMidiBridge();
        
        bridgeSend.getFile = function(path_p) {
			if (path_p == "foo") {
				return "OllehCasklakldmcömpöosdpo sdvsdöv sdvsd vsdvgeg etrhtrhrt ret";
			}
			if (path_p == "bar") {
				return "Hello sdv wrpv wprvmevmkefv ö löevö wöorvömklfsdvnlsd dvsdvsdvsdvsssvsd";
			}						
		}
		
        // Bridge for receiving data
        const bridgeReceive = new JsMidiBridge();

		let finishCalls = [];
		
		bridgeReceive.onReceiveFinish = function(data) {
			finishCalls.push(data.data);
		}
		
		let messagesSentSend = [];
		bridgeSend.sendSysex = function(manufacturerId, data) {
			messagesSentSend.push({
                manufacturerId: manufacturerId,
                data: data
            });
		}

        let messagesSentReceive = [];
		bridgeReceive.sendSysex = function(manufacturerId, data) {
			messagesSentReceive.push({
                manufacturerId: manufacturerId,
                data: data
            });
		}	

        // Get a request message for the path from the receiving side
        await bridgeReceive.request("foo", 3);
        const msgRequestFoo = messagesSentReceive[0];
        messagesSentReceive = [];

        // Get a request message for the path from the receiving side
        await bridgeReceive.request("bar", 4);
        const msgRequestBar = messagesSentReceive[0];
        messagesSentReceive = [];

        // Let the bridge receive a request message, to trigger it sending a file
        expect(await bridgeSend.receive(msgRequestFoo)).toBe(true);

        // Receive start message
        const startMsg1 = messagesSentSend.shift();
        const dataMsg1 = messagesSentSend.shift();
        expect(await bridgeReceive.receive(startMsg1)).toBe(true);
        expect(messagesSentReceive.length).toBe(0);

        // Receive data message
        expect(await bridgeReceive.receive(dataMsg1)).toBe(true);

        // Get next data message
        expect(messagesSentReceive.length).toBe(1);
        const ackMsg1 = messagesSentReceive.shift();
        
        // Let pass some virtual time so the first transmission will timeout
        jasmine.clock().mockDate(new Date(1000 * 10));
        
        // Let the sending bridge receive a request message, to trigger it sending a file
        expect(await bridgeSend.receive(msgRequestBar)).toBe(true);
        
        messagesSentSend = [];
        expect(await bridgeSend.receive(ackMsg1)).toBe(true);

        await this.#evaluateError(bridgeSend, messagesSentSend, ["transmission", "not found"]);
    }


//###############################################################################################


    /**
	 * Helper to check error messages
	 */
    async #evaluateError(bridge, messagesSent, tokens) {
        expect(messagesSent.length).toBe(2);

        const otherBridge = new JsMidiBridge(); 

		let called = false;

		otherBridge.onError = function(message) {
            for(const token of tokens) {
                expect(message).toContain(token);
            }
			
			called = true					
		}

        let messagesSentOther = [];
		otherBridge.sendSysex = function(manufacturerId, data) {
			messagesSentOther.push({
                manufacturerId: manufacturerId,
                data: data
            });	
		}
                
        while(true) {
            if (messagesSent.length == 0) {
                break;
            }                

            const msg = messagesSent.shift();
            
            expect(await otherBridge.receive(msg)).toBe(true);

            while (messagesSentOther.length > 0) {
                await bridge.receive(messagesSentOther.shift());
            }
        }

        expect(called).toBe(true);
    }
}

File: python/test/pymidibridge/test_protocol_timeout.py
import sys
import unittest
from unittest.mock import patch   # Necessary workaround! Needs to be separated.

from .mocks import *

with patch.dict(sys.modules, {
    "time": MockTime
}):
    from lib.pymidibridge.PyMidiBridge import *
    

class TestProtocolTimeout(unittest.TestCase):

    def test_receive_timeout(self):
        MockTime.monotonic_return = 1

        # Bridge for sending data
        midi_send = MockMidiSender()
        storage_send = MockStorageProvider()        

        storage_send.read_data = {
            "foo": "OllehCasklakldmcömpöosdpo sdvsdöv sdvsd vsdvgeg etrhtrhrt ret",
            "bar": "Hello sdv wrpv wprvmevmkefv ö löevö wöorvömklfsdvnlsd dvsdvsdvsdvsssvsd"
        }

        storage_send.outputs_size = {
            "foo": len(storage_send.read_data["foo"]),
            "bar": len(storage_send.read_data["bar"])
        }

        bridge_send = PyMidiBridge(
            midi = midi_send,
            storage_factory = get_mock_storage_factory(storage_send),
            event_handler = MockEventHandler()
        )

        # Bridge for receiving data
        midi_receive = MockMidiSender()
        storage_receive = MockStorageProvider()        

        bridge_receive = PyMidiBridge(
            midi = midi_receive,
            storage_factory = get_mock_storage_factory(storage_receive),
            event_handler = MockEventHandler()
        )

        # Get a request message for the path from the receiving side
        bridge_receive.request("foo", 3)
        msg_request_foo = midi_receive.messages_sent[0]
        midi_receive.messages_sent = []

        # Get a request message for the path from the receiving side
        bridge_receive.request("bar", 4)
        msg_request_bar = midi_receive.messages_sent[0]
        midi_receive.messages_sent = []

        # Let the sending bridge receive a request message, to trigger it sending a file
        self.assertEqual(bridge_send.receive(msg_request_foo), True)        
        
        # Receive start message 
        start_msg_1 = midi_send.messages_sent.pop(0)
        data_msg_1 = midi_send.messages_sent.pop(0)
        self.assertEqual(bridge_receive.receive(start_msg_1), True)
        self.assertEqual(len(midi_receive.messages_sent), 0)

        # Receive data message
        self.assertEqual(bridge_receive.receive(data_msg_1), True)

        # Get next data message
        self.assertEqual(len(midi_receive.messages_sent), 1)
        ack_msg_1 = midi_receive.messages_sent.pop(0)
        self.assertEqual(bridge_send.receive(ack_msg_1), True)
        data_msg_1b = midi_send.messages_sent.pop(0)
        
        # Let pass some virtual time so the first transmission will timeout
        MockTime.monotonic_return += 10
        
        # Let the sending bridge receive a request message, to trigger it sending a file
        self.assertEqual(bridge_send.receive(msg_request_bar), True)        
        
        # Receive start message 
        start_msg_2 = midi_send.messages_sent.pop(0)
        data_msg_2 = midi_send.messages_sent.pop(0)
        self.assertEqual(bridge_receive.receive(start_msg_2), True)
        self.assertEqual(len(midi_receive.messages_sent), 0)

        # Receive data message 2 (must raise "Transmission not found")
        self.assertEqual(bridge_receive.receive(data_msg_1b), True)

        self._evaluate_error(bridge_receive, midi_receive, ["transmission", "not found"])


    def test_send_timeout(self):
        MockTime.monotonic_return = 1

        # Bridge for sending data
        midi_send = MockMidiSender()
        storage_send = MockStorageProvider()        

        storage_send.read_data = {
            "foo": "OllehCasklakldmcömpöosdpo sdvsdöv sdvsd vsdvgeg etrhtrhrt ret",
            "bar": "Hello sdv wrpv wprvmevmkefv ö löevö wöorvömklfsdvnlsd dvsdvsdvsdvsssvsd"
        }

        storage_send.outputs_size = {
            "foo": len(storage_send.read_data["foo"]),
            "bar": len(storage_send.read_data["bar"])
        }

        bridge_send = PyMidiBridge(
            midi = midi_send,
            storage_factory = get_mock_storage_factory(storage_send),
            event_handler = MockEventHandler()
        )

        # Bridge for receiving data
        midi_receive = MockMidiSender()
        storage_receive = MockStorageProvider()        

        bridge_receive = PyMidiBridge(
            midi = midi_receive,
            storage_factory = get_mock_storage_factory(storage_receive),
            event_handler = MockEventHandler()
        )

        # Get a request message for the path from the receiving side
        bridge_receive.request("foo", 3)
        msg_request_foo = midi_receive.messages_sent[0]
        midi_receive.messages_sent = []

        # Get a request message for the path from the receiving side
        bridge_receive.request("bar", 4)
        msg_request_bar = midi_receive.messages_sent[0]
        midi_receive.messages_sent = []

        # Let the sending bridge receive a request message, to trigger it sending a file
        self.assertEqual(bridge_send.receive(msg_request_foo), True)        
        
        # Receive start message 
        start_msg_1 = midi_send.messages_sent.pop(0)
        data_msg_1 = midi_send.messages_sent.pop(0)
        self.assertEqual(bridge_receive.receive(start_msg_1), True)
        self.assertEqual(len(midi_receive.messages_sent), 0)

        # Receive data message
        self.assertEqual(bridge_receive.receive(data_msg_1), True)

        # Get next data message
        self.assertEqual(len(midi_receive.messages_sent), 1)
        ack_msg_1 = midi_receive.messages_sent.pop(0)

        # Let pass some virtual time so the first transmission will timeout
        MockTime.monotonic_return += 10
        
        # Let the sending bridge receive a request message, to trigger it sending a file
        self.assertEqual(bridge_send.receive(msg_request_bar), True)        
        
        midi_send.messages_sent = []
        self.assertEqual(bridge_send.receive(ack_msg_1), True)

        self._evaluate_error(bridge_send, midi_send, ["transmission", "not found"])

        
############################################################################################################


    # Helper to check error messages.
    def _evaluate_error(self, bridge, midi, tokens):
        self.assertEqual(len(midi.messages_sent), 2)

        events = MockEventHandler()
        other_midi = MockMidiSender()
        other_bridge = PyMidiBridge(
            midi = other_midi, 
            storage_factory = get_mock_storage_factory(),
            event_handler = events
        )        

        events.last_error = None
        
        while True:
            if not midi.messages_sent:
                break

            msg = midi.messages_sent.pop(0)
            
            self.assertEqual(other_bridge.receive(msg), True)

            while other_midi.messages_sent:
                bridge.receive(other_midi.messages_sent.pop(0))
                
        self.assertIsNotNone(events.last_error, "Exception not thrown: " + repr(tokens))

        for token in tokens:
            self.assertIn(token, events.last_error)

